File: flutter/lib/features/champion/data/repository/patch_version_repository_impl.dart
import 'package:dio/dio.dart';
import 'package:lol_champion/core/error/app_exception.dart';
import 'package:lol_champion/features/champion/data/datasource/patch_version_local_data_source.dart';
import 'package:lol_champion/features/champion/data/datasource/patch_version_remote_data_source.dart';
import 'package:lol_champion/features/champion/domain/repository/patch_version_repository.dart';

class PatchVersionRepositoryImpl implements PatchVersionRepository {
  const PatchVersionRepositoryImpl(this._remote, this._local);

  final PatchVersionRemoteDataSource _remote;
  final PatchVersionLocalDataSource _local;

  @override
  Future<String> getPatchVersion() => _local.read();

  @override
  Future<String> refreshLatestPatchVersion() async {
    try {
      final latestVersion = await _remote.getLatestPatchVersion();
      if (latestVersion == null) {
        return _local.read();
      }
      await _local.write(latestVersion);
      return latestVersion;
    } on DioException catch (error) {
      throw NetworkException(error);
    } on Object catch (error) {
      throw DataException(error);
    }
  }
}

File: react/src/features/champion/data/repositories/patch-version-repository.impl.ts
import { appConfig } from '@/core/config/app-config';
import type { PatchVersionLocalDataSource } from '@/features/champion/data/data-sources/patch-version-local-data-source';
import type { PatchVersionRemoteDataSource } from '@/features/champion/data/data-sources/patch-version-remote-data-source';
import type { PatchVersionRepository } from '@/features/champion/domain/repositories/patch-version-repository';

export class ResilientPatchVersionRepository implements PatchVersionRepository {
  constructor(
    private readonly remoteDataSource: PatchVersionRemoteDataSource,
    private readonly localDataSource: PatchVersionLocalDataSource,
  ) {}

  async resolveLatestVersion(): Promise<string> {
    try {
      const version = await withTimeout(
        this.remoteDataSource.fetchLatestVersion(),
        appConfig.startupTimeoutMs,
      );
      try {
        this.localDataSource.writeVersion(version);
      } catch {
        // Browser storage can be unavailable; the fresh version is still valid.
      }
      return version;
    } catch {
      try {
        return this.localDataSource.readVersion() || appConfig.fallbackPatchVersion;
      } catch {
        return appConfig.fallbackPatchVersion;
      }
    }
  }
}

async function withTimeout<T>(promise: Promise<T>, timeoutMs: number): Promise<T> {
  let timeoutId: ReturnType<typeof setTimeout> | undefined;
  const timeout = new Promise<never>((_, reject) => {
    timeoutId = setTimeout(() => reject(new Error('Patch version request timed out')), timeoutMs);
  });

  try {
    return await Promise.race([promise, timeout]);
  } finally {
    clearTimeout(timeoutId);
  }
}
